### PR TITLE
Optimize Poincare model training

### PR DESCRIPTION
**Pull request by koiizukag**
Reduce computational complexity O(N^2) => O(N).
This part is the bottleneck of the overall training phase.

### CODE REVIEW

File: gensim/models/poincare.py [hunk 1]
```diff
@@ -567,10 +567,12 @@ def _handle_duplicates(vector_updates, node_indices):
 
         """
         counts = Counter(node_indices)
+        node_dict = defaultdict(list)
+        [node_dict[node_index].append(i) for i, node_index in enumerate(node_indices)]
```
Comment (piskvorky) [inline gensim/models/poincare.py]: Rewrite as loop (unpythonic construct).

**Comment by koiizukag** [comment]
Thanks @piskvorky . 

I found this insufficient implementation via python profiler.
- https://github.com/rkern/line_profiler
- https://github.com/what-studio/profiling

If we use multiprocessing in independent for loop, we can make this model faster.
ex: https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/poincare.py#L545
Also, it is better to use Cython like other gensim models.

**Comment by mpenkov** [comment]
Thank you for your contribution @koiizukag.

@piskvorky This looks good to merge to me. Is there anything left on this PR that you want to do before we merge this?

**Comment by piskvorky** [comment]
LGTM. My only question was around fixing other similar issues in Poincare, since we (@koiizukag ) are at it. But nothing blocking.

**Pull request merged**